records_mover/url/optimizer/__init__.py

```diff
@@ -0,0 +1,113 @@
+from contextlib import contextmanager
+from records_mover.url.base import BaseDirectoryUrl
+import logging
+from .gcp_data_transfer_service import GcpDataTransferService
+from .s3_copy_via_aws_cli import S3CopyViaAwsCli
+from typing import Iterator, Tuple
+
+
+logger = logging.getLogger(__name__)
+
+
+class CopyOptimizer:
+    """When Records Mover needs to move data between different URL
+    schemes, the default way (a single-threaded downloading of bytes
+    from the source and then reuploading of bytes to the target) can
+    be very slow for large directories.
+
+    This class can use faster means (ideally a cloud-based service
+    that can copy things at scale, or at least vendor-optimized
+    multi-threaded upload/download code) when they are available.
+
+    This depends on the specifics of the copy, and as such return
+    values must be checked to determine if the copy was able to be
+    performed.
+    """
+
+    def __init__(self) -> None:
+        self._gcp_data_transfer = GcpDataTransferService()
+        self._s3_copy_via_aws_cli = S3CopyViaAwsCli()
+
+    def copy(self,
+             loc: BaseDirectoryUrl,
+             other_loc: BaseDirectoryUrl) -> bool:
+        """Copy from one directory to another using optimized means.
+
+        :return: True if the copy was performed, or False if no
+        optimized means were possible."""
+        # TODO: Can I move some of this logic into the other classes?
```

Comment:
Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker